File: datasets/binary_search_tree/type_3_structural.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BST:
    def __init__(self):
        self.root = None

    def add(self, node, value):
        if node is None:
            return TreeNode(value)

        if value < node.value:
            node.left = self.add(node.left, value)
        else:
            node.right = self.add(node.right, value)

        return node

    def find(self, node, value):
        if node is None or node.value == value:
            return node

        if value < node.value:
            return self.find(node.left, value)

        return self.find(node.right, value)

File: server/datasets/controlled_benchmarks/java/dsa_clone_benchmarks/binary_search_tree/type_3_structural.java
class TreeNode {
    int value;
    TreeNode left, right;

    TreeNode(int value) {
        this.value = value;
        left = right = null;
    }
}

class BST {
    TreeNode root;

    public TreeNode add(TreeNode node, int value) {
        if (node == null) {
            return new TreeNode(value);
        }

        if (value < node.value) {
            node.left = add(node.left, value);
        } else {
            node.right = add(node.right, value);
        }

        return node;
    }

    public TreeNode find(TreeNode node, int value) {
        if (node == null || node.value == value) {
            return node;
        }

        if (value < node.value) {
            return find(node.left, value);
        }

        return find(node.right, value);
    }
}